### PR TITLE
[consensus] Add Commit Observer

### PR DESCRIPTION
**Pull request by arun-koshy**
Copying comment from PR

```
/// Role of CommitObserver
/// - Called by core when block is created via try_new_block or force_new_block
/// and try_commit returns newly comitted leaders
/// - The newly committed leaders are sent to commit observer and then commit observer
/// gets subdags for each leader via the commit interpreter (linearizer)
/// - The committed subdags are sent as consensus output via an unbounded tokio channel.
/// No back pressure mechanism is needed as backpressure is handled as input into
/// consenus.
/// - Last commit index is persisted in store as soon as the subdag is sent to
/// output channel.
/// - When CommitObserver is initialized a last received commit index can be used
/// to ensure any missing commits are re-sent.
```

### CODE REVIEW

File: consensus/core/src/linearizer.rs [hunk 1]
```diff
@@ -177,9 +178,13 @@ impl Linearizer {
...
     // This function should be called whenever a new commit is observed. This will
-    // iterate over the sequence of committed leaders and produce a list of sub-dags.
-    pub fn handle_commit(&mut self, committed_leaders: Vec<VerifiedBlock>) -> Vec<CommittedSubDag> {
-        let mut committed_sub_dags = vec![];
+    // iterate over the sequence of committed leaders and produce a list of commit
+    // data & committed sub-dags.
+    pub fn handle_commit(
+        &mut self,
+        committed_leaders: Vec<VerifiedBlock>,
+    ) -> Vec<(Commit, CommittedSubDag)> {
```
Comment (mwtian) [inline consensus/core/src/linearizer.rs:186]: If Commit is returned only for persistence, we can revert the change and ignore the persistence logic in this PR. Commit can be buffered in DagState, which needs access to the last Commit as well.

File: consensus/core/src/commit_observer.rs [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+        // Commit all of the leaders and "receive" the subdags as the consumer of
+        // the consensus output channel.
+        let expected_last_received_index = 3;
+        let commits = observer.handle_commit(leaders.clone());
+
+        // Check commits sent over consensus output channel is accurate
+        let mut received_subdag_index = 0;
+        while let Ok(subdag) = receiver.try_recv() {
+            tracing::info!("Received {subdag}");
+            assert_eq!(subdag.leader, commits[received_subdag_index].leader);
```
Comment (mwtian) [inline consensus/core/src/commit_observer.rs:449]: Can we `#[derive(PartialEq)]` on `CommittedSubDag` and check equivalence?

File: consensus/core/src/linearizer.rs (hunk 1 above)
Comment (mwtian) [reply consensus/core/src/linearizer.rs]: This sounds good for now.

**Review by mwtian — APPROVED** [review]

File: consensus/core/src/commit_observer.rs [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+        let commits = self.commit_interpreter.handle_commit(committed_leaders);
+        let mut sent_sub_dags = vec![];
+        let mut sent_commit_data = vec![];
+
+        for (commit_data, committed_sub_dag) in commits.into_iter() {
+            if let Err(err) = self.sender.send(committed_sub_dag.clone()) {
+                tracing::error!(
+                    "Failed to send committed sub-dag, probably due to shutdown: {err:?}"
+                );
+                break;
```
[the remark this answers is not in the document]
Comment (arun-koshy) [reply consensus/core/src/commit_observer.rs]: For now will break as soon as a sender error is hit and left a todo to revisit sending a signal back to core.

**Review by asonnino — APPROVED** [review]
sorry for the delay, lgtm

File: consensus/core/src/commit_observer.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (akichidis) [reply consensus/core/src/commit_observer.rs:75]: I see all the concerns here. I believe that this way for now is definitely ok. But yes we might want to revisit to ensure that Core is shutdown as safe as possible. I don't see how this channel can deal with transient errors, but in the end we should probably adopt whatever is the 100% safe.

File: consensus/core/src/dag_state.rs [hunk 4]
```diff
@@ -234,7 +234,25 @@ impl DagState {
...
+        let last_commit = commits.last().unwrap().clone();
+        self.store
+            .write(committed_blocks, commits)
+            .expect("Writing commits to store should not fail");
+        self.set_last_commit(last_commit);
+    }
+
     pub(crate) fn set_last_commit(&mut self, commit: Commit) {
+        if let Some(last_commit) = &self.last_commit {
+            assert!(commit.index >= last_commit.index);
```
Comment (akichidis) [inline consensus/core/src/dag_state.rs:254]: In which case we expect to be `commit.index == last_commit.index` ? Given the recovery path and the common case I believe we should always move forward , right?

File: consensus/core/src/commit_observer.rs [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+        // Check commits sent over consensus output channel is accurate
+        let mut processed_subdag_index = 0;
+        while let Ok(subdag) = receiver.try_recv() {
+            tracing::info!("Processed {subdag}");
+            assert_eq!(subdag, commits[processed_subdag_index]);
+            processed_subdag_index = subdag.commit_index as usize;
+            if processed_subdag_index == expected_last_processed_index {
+                break;
+            }
+        }
```
Comment (akichidis) [inline consensus/core/src/commit_observer.rs:431]: I believe might need to assert after the loop that the `processed_subdag_index` has been updated to the latest expected value as there is the case where the channel is from the beginning empty. In this scenario the `receiver.try_recv` will just return an `Err` for `Empty` channel directly and the loop will never process. I see that we do verify the last stored commit , but I think it would be good to make sure that the loop did run indeed

**Review by akichidis — APPROVED** [review]
💯

File: consensus/core/src/dag_state.rs (hunk 4 above)
Comment (arun-koshy) [reply consensus/core/src/dag_state.rs:254]: It is very subtle, but the way it works currently is that the linearizer batch committed subdags to write to store but it will update the last commit in dag state on each collected subdag so that the next collected subdag can use that as a reference to ensure no duplicate blocks are collected. So once the batch of subdags are eventually written to store the last commit in dagstate will be equal to the last one already set by linearizer during collection.

Can follow up in the next PR if there are concerns around this.
